Add tDL prediction verbalizer and additional supports

moreshud opened the pull request:
- Added `verbalize_learner_prediction` for verbalizing tDL prediction with `verbalize` flag
- Updated the `main.py` and `executor.py` for nces and tdl
- Updated the `concept_learning_cv_evaluation.py` and `README.md` for either single, selected, or all learners support

moreshud committed:
chores: add results to gitignore

moreshud committed:
fixes: update nces in executor with chores

moreshud committed:
fix: include tdl in executor

moreshud committed:
fix: update for leaner choices in cv evaluation

moreshud committed:
chores:  update evolearner leaner types

moreshud committed:
feat: add verbalize for tdl

moreshud committed:
fix: add cv leaner_types in README and learner prediction verbalizer

moreshud committed:
fix: add openai to setup.py

moreshud committed:
fix: add openai to setup.py

moreshud committed:
fix: add openai to setup.py

Demirrr commented:
@moreshud could you please use the cursor/windsurf or copilot to 

1. Add docstrings
2. Document the computation
3. Type information and
4. Add mock unit tests and regression tests

@alkidbaci could you take over the reviewing?

moreshud committed:
fixes: added docstring and unittest

moreshud commented:
> @moreshud could you please use the cursor/windsurf or copilot to
> 
> 1. Add docstrings
> 2. Document the computation
> 3. Type information and
> 4. Add mock unit tests and regression tests
> 
> @alkidbaci could you take over the reviewing?

- Added verbalize_learner_prediction for verbalizing tDL prediction with verbalize flag
  - Upon intializing the tDL with verbalize = True, it allows the call `verbalize_learner_prediction(self.disjunction_of_conjunctive_concepts)` within tDL class which returns natural language interpretation of the argument using `LLMVerbalizer()` (with default settings or with private access credentials)

- Updated the `main.py` and `executor.py` for nces and tdl
  - Added tdl to `args.model` choices and updated `executor.py` with tdl and nces
    - `python3 main.py --model tdl --knowledge_base_path ../data/KGs/Family/family-benchmark_rich_background.owl --path_learning_problem ../data/LPs/Family/lps.json`
    - `python3 main.py --model nces --knowledge_base_path ../data/KGs/Family/family-benchmark_rich_background.owl --path_learning_problem ../data/LPs/Family/lps.json --path_of_nces_embeddings ../data/NCESData/family/embeddings/DeCaL_entity_embeddings.csv --path_of_nces_trained_models ../data/NCESData/family/trained_models`
  - Added automatic saving of `executor.py` learner data with the convention `[kb_file_name]_[model_type].csv` in the `results` directory (added to gitignore) using the `get_file_base_name` and `prepare_output_path`
  
- Refactor the 'concept_learning_cv_evaluation.py' for either single, selected, or all learners support and updated README.md with this.
  ```
    # run all learners for cross-validation (with or without --learner_types)
    - python3 examples/concept_learning_cv_evaluation.py --folds 3  --kb ../data/KGs/Family/family-benchmark_rich_background.owl --lps ../data/LPs/Family/lps.json --path_of_nces_embeddings ../data/NCESData/family/embeddings/DeCaL_entity_embeddings.csv --path_of_nces_trained_models ../data/NCESData/family/trained_models

    - python3 examples/concept_learning_cv_evaluation.py --learner_types --folds 3  --kb ../data/KGs/Family/family-benchmark_rich_background.owl --lps ../data/LPs/Family/lps.json --path_of_nces_embeddings ../data/NCESData/family/embeddings/DeCaL_entity_embeddings.csv --path_of_nces_trained_models ../data/NCESData/family/trained_models

    # run selected learners for cross-validation
    - python3 examples/concept_learning_cv_evaluation.py --learner_types ocel tdl nces --folds 3  --kb ../data/KGs/Family/family-benchmark_rich_background.owl --lps ../data/LPs/Family/lps.json --path_of_nces_embeddings ../data/NCESData/family/embeddings/DeCaL_entity_embeddings.csv --path_of_nces_trained_models ../data/NCESData/family/trained_models

    # run single learner for cross-validation
    - python3 examples/concept_learning_cv_evaluation.py --learner_types tdl --folds 3  --kb ../data/KGs/Family/family-benchmark_rich_background.owl --lps ../data/LPs/Family/lps.json --path_of_nces_embeddings ../data/NCESData/family/embeddings/DeCaL_entity_embeddings.csv --path_of_nces_trained_models ../data/NCESData/family/trained_models
  ```

Demirrr commented:
Too many files are updated in a single PR. Please refrain from this. @alkidbaci please proceed with the review

alkidbaci commented on ontolearn/verbalizer.py:
```diff
@@ -23,27 +23,69 @@
\ No newline at end of file
+            return f"No verbalization. Error at HTTP connection"
+        
+def verbalize_learner_prediction(prediction: Union[str, object] = None) -> None:
```
The number of predictions to be printed can be set via a method argument

moreshud replied:
I have added the num_prediction with default value of 3

alkidbaci replied:
please also update the docstring in accordance

moreshud committed:
fix: add num_prediction to verbalize_learner_prediction

moreshud committed:
chore: complete docstring for verbalizer num_predictions

alkidbaci committed:
Merge remote-tracking branch 'origin/develop' into tdl

alkidbaci reviewed (approved):
We can merge once the the checks pass successfully

alkidbaci commented:
Thanks for the PR @moreshud
